### PR TITLE
#5666 Add initial support for Thunderbird (improved setup flow)

### LINKED ISSUE #5665: [thunderbird] Recognize thunderbird environment and use alternative to gmail element replacer class

### LINKED ISSUE #5666: [thunderbird] Add compose button into Thunderbird UI

### PR DESCRIPTION
**Pull request by martgil**
This PR adds initial support for the Thunderbird desktop email client and the specific goal is to improve the set up flow accordingly.

close #5665
issue #5666

----------------------------------

**Tests** _(delete all except exactly one)_:
- Difficult to test - (changes are not readily completed for tests)

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

**Comment by tomholub** [comment]
How is this looking so far? This looks like it could close https://github.com/FlowCrypt/flowcrypt-browser/issues/5665 nearly as is. Then you could work on the followup issues in separate PRs. It's always better to make PRs smaller if possible to make progress easier.

One thing I noticed is that the `content_scripts/webmail` folder already on master is quite disorganised and not so easy to work on to add Thunderbird (or any other webmail) support. I country try and refactor it a little bit myself, and that could make it easier for you.

**Comment by tomholub** [comment]
When https://github.com/FlowCrypt/flowcrypt-browser/pull/5748 is merged, you can follow https://github.com/FlowCrypt/flowcrypt-browser/pull/5748#issuecomment-2135555543 to close #5665 with this PR, then start working on #5666 in next issue. Let's address these one by one instead of trying to do all at once.

**Comment by martgil** [comment]
I understood completely, Tom. I'll be following your advice instead of making them work all at once.

**Comment by martgil** [comment]
Hi @sosnovsky - This one is ready for review. Thank you!

**Comment by martgil** [comment]
Hello @sosnovsky - For me everything works well. I'll be sharing a video screen capture and send it to you over secure email. Nothing much to test here aside from customize descriptions as if it were a Thunderbird extension. I have also figured small things about the compose button implementation that still requires thorough planning -- let's discuss it on the original issue after this one.

**Comment by martgil** [comment]
> I installed FlowCrypt extension in Thunderbird, configured it and now can see FlowCrypt button on toolbar.
But when opening Encrypted inbox, it loads just empty page.

I have re-checked it and it looks like it happens because the "chrome" object was tried to be access when it fact it should be the "browser" object for firefox extensions. I'll review again and hoping it would be fixed.

**Comment by martgil** [comment]
I noticed that when accessing the extension not in the toolbar but in the Add-ons Manager of the Thunderbird mail settings and clicking the 🔧 icon, every extension page works fine. I'm investigating the issue further to develop a better hypothesis to share.

**Comment by martgil** [comment]
Hi @sosnovsky - I have fixed the issue you have shared. You should now be able to access the FlowCrypt Settings and FlowCrypt inbox directly through the default_popup.htm on the Thunderbird toolbar. Thank you!

**Comment by martgil** [comment]
As for testing, undergo normal setup using the Thunderbird extension build. Once you have finished the setup, try again accessing the FlowCrypt settings/inbox from the Thunderbird toolbar and it should work well.

**Comment by martgil** [comment]
@sosnovsky This one is ready for review. thank you!

File: extension/js/common/browser/browser.ts [hunk 1]
```diff
@@ -55,13 +56,14 @@ export class Browser {
   public static openSettingsPage = async (path = 'index.htm', acctEmail?: string, page = '', rawPageUrlParams?: Dict<UrlParam>, addNewAcct = false) => {
     const basePath = chrome.runtime.getURL(`chrome/settings/${path}`);
     const pageUrlParams = rawPageUrlParams ? JSON.stringify(rawPageUrlParams) : undefined;
+    const preferedLinkOpener = Catch.browser().name === 'thunderbird' ? BgUtils : Browser;
```
Comment (sosnovsky) [inline extension/js/common/browser/browser.ts:59]: I think it'll be simpler to add this check to `Browser.openExtensionTab`, so it'll look like this:
```ts
public static openExtensionTab = (url: string) => {
  if (Catch.browser().name === 'thunderbird') {
    BgUtils.openExtensionTab(url)
  } else {
    ...
  }
}
```

File: tooling/build-types-and-manifests.ts [hunk 2]
```diff
@@ -49,6 +49,43 @@ addManifest('firefox-consumer', manifest => {
   delete manifest.minimum_chrome_version;
 });
 
+addManifest('thunderbird-consumer', manifest => {
```
Comment (sosnovsky) [inline tooling/build-types-and-manifests.ts:52]: are there many differences with `firefox-consumer` manifest?
probably it'll be simpler to reuse manifest generated for firefox and just update fields which are different

**Review by sosnovsky — COMMENTED** [review]
It works well for me now 👍

File: extension/js/common/browser/browser.ts (hunk 1 above)
Comment (martgil) [reply extension/js/common/browser/browser.ts]: Thanks Roma - your suggestion is way better. Will update the changes.

File: tooling/build-types-and-manifests.ts (hunk 2 above)
Comment (martgil) [reply tooling/build-types-and-manifests.ts]: Thank you, Roma - I missed that. I have update the code to reuse a specific manifest file and update the fields that differs. thanks!

**Comment by martgil** [comment]
Hello @sosnovsky - I have added the required changes, tested it and confirmed working well. Thank you!

File: extension/js/content_scripts/webmail/thunderbird/thunderbird-webmail-startup.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    clientConfiguration: ClientConfiguration,
+    injector: Injector,
+    notifications: Notifications
+    // factory: XssSafeFactory, // todo in another issue
+    // relayManager: RelayManager // todo in another issue
+  ) => {
+    // injector.btns(); // todo in another issue - add compose button
+    this.replacer.runIntervalFunctionsPeriodically();
+    await notifications.showInitial(acctEmail);
+    await notifications.show(
```
Comment (sosnovsky) [inline extension/js/content_scripts/webmail/thunderbird/thunderbird-webmail-startup.ts:34]: no need for `await` here

**Review by sosnovsky — COMMENTED** [review]
Well done 👍

File: extension/js/content_scripts/webmail/thunderbird/thunderbird-webmail-startup.ts (hunk 3 above)
Comment (martgil) [reply extension/js/content_scripts/webmail/thunderbird/thunderbird-webmail-startup.ts]: updated - thank you!

**Review by sosnovsky — APPROVED** [review]
All good now, thanks! I've just updated issue description to link correct issues, as initially this PR would close `Add compose button into Thunderbird UI` issue

**Comment by martgil** [comment]
Thank you @sosnovsky - I can now drop some of the few questions that needs to be answer regarding adding compose button into thunderbird UI on its dedicated GitHub issue. Thank you for your help.

**Pull request merged**